### ISSUE TITLE
More options in events filtering 

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#2853]
## 🚀 Feature

The idea is coming from this discussion: https://github.com/Project-MONAI/MONAI/discussions/5958#discussioncomment-4920024

We would like to support the following use-cases:
1) `before/after` can work together with `every`
```
event = Events.ITERATION_COMPLETED(after=9, before=21, every=5)
# execute for iterations [10, 15, 20]
```
2) `once` can accept list of values
```
event = Events.ITERATION_COMPLETED(once=[5, 8, 23])
# execute once for iterations 5, 8 and 23
```

### How to help with this issue

Events filtering logic is coded here: https://github.com/pytorch/ignite/blob/5d344a06e32d2601517eccce007738d29cb703a6/ignite/engine/events.py#L50-L119

We would like to update the following functions:

https://github.com/pytorch/ignite/blob/5d344a06e32d2601517eccce007738d29cb703a6/ignite/engine/events.py#L110

and 

https://github.com/pytorch/ignite/blob/5d344a06e32d2601517eccce007738d29cb703a6/ignite/engine/events.py#L113

such that we could cover proposed use-cases. Finally, we need to write few tests here: https://github.com/pytorch/ignite/blob/master/tests/ignite/engine/test_custom_events.py

**Comment by DeepC004** [comment]
@vfdev-5 I would like to work on the issue. Please assign it to me.

**Comment by DeepC004** [comment]
I can take up implementation of `once` as referred in the example above. From the discussion [Project-MONAI/MONAI#5958 (reply in thread)](https://github.com/Project-MONAI/MONAI/discussions/5958#discussioncomment-4920024), `once` can be of type `int | list[int]` so I need to add the functionality of `once` being of type `list[int]` in the current implementation?

**Comment by vfdev-5** [comment]
@jak0bw we would like to have your feedback on the behaviour in case of 
```python
event = Events.ITERATION_COMPLETED(after=8, before=25, every=5)
```
It can be either: 1) execute for iterations [10, 15, 20] => every 5 iters since 0
or it can be 2) execute for iterations [9, 14, 19, 24] => every 5 iters since `after`

Basically, the question is, should we do
```python
before = 25
after = 8
every = 5

for event in range(50):
    if after < event < before and (event - after - 1) % every == 0:
        print(event)
```
Output: 9 14 19 24

or 
```python
before = 25
after = 8
every = 5

for event in range(50):
    if after < event < before and event % every == 0:
        print(event)
```
Output: 10 15 20

What do you think ?

**Comment by theory-in-progress** [comment]
Exactly! Well, according to me, the `after` and `before` parameters specify a range of iterations during which the event should be triggered, and the `every` parameter specifies that the event should only be triggered every `every` iterations.

Suppose for `event = Events.ITERATION_COMPLETED(after=8, before=25, every=5)`, the range of iterations is from `8` to `25`, and the event should only be triggered every `5` iterations.
Hence printing: `[9 14 19 24]`

**Comment by jak0bw** [comment]
I thought about this as well when stating the idea. Basically, I agree with @theory-in-progress. I think this way
 > ```python
> before = 25
> after = 8
> every = 5
> 
> for event in range(50):
>     if after < event < before and (event - after - 1) % every == 0:
>         print(event)
> ```
> 
> Output: 9 14 19 24

is more flexible as it can do both `[9 14 19 24]` and for a different parametrization `before=9 after=21 ever=5` we can get `10 15 20.` This is not possible if we use the other way, as it would always be confined to multiples of `every`.

**Comment by jak0bw** [comment]
Just had one more thought on this topic that might be interesting.

Perhaps it's desirable to change before and after to first and last. Then, the parametrization of first=10, last=20, every=5 would result in [10,15,20] which in my opinion looks just a little bit better than after=9, before=21, every=5...

But this is of course not really a breaking feature 🤷🏻‍♂️

**Comment by vfdev-5** [comment]
> Perhaps it's desirable to change before and after to first and last.

As we can also use `before` and `after` by defining only one of them: `ITERATION_STARTED(after=10)` which is looks ok then if we replace `after` by `first` it would seem a bit strange I think: `ITERATION_STARTED(first=10)`. 

By the way, right now before and after can only define an interval like [after, before] and event filtering wont work if we would like to pick events inside `[0, before] or `[after, inf]`. I wonder if we should also support that ?

**Comment by jak0bw** [comment]
> By the way, right now before and after can only define an interval like `[after, before]` and event filtering wont work if we would like to pick events inside `[0, before]` or `[after, inf]`. I wonder if we should also support that ?

I think it should be possible to only give every possible subset of  `after, every, before` and omit the rest. With the implication being that the default values are logically `after=-1, every=1, before=+inf` (Even if this is solved differently in code).

> As we can also use `before` and `after` by defining only one of them: `ITERATION_STARTED(after=10)` which is looks ok then if we replace `after` by `first` it would seem a bit strange I think: `ITERATION_STARTED(first=10)`.

I am not quite sure if I agree with you here, as in my opinion `first=10` implies every iteration from `10` to `inf `in the same way as `after=9` does  (and the same argument for `last=20` <-> `before=21`). But as I said this might be personal preference, just wanted to state the idea 🤔

**Comment by theory-in-progress** [comment]
> By the way, right now before and after can only define an interval like `[after, before]` and event filtering wont work if we would like to pick events inside `[0, before]` or `[after, inf]`. I wonder if we should also support that ?

Please Correct me If I am Wrong, but I Think we already support that? We set default values of `after` and `before` When we enter the function, both of them are optional:

https://github.com/pytorch/ignite/blob/89d29303bcc11ce85ec5e010f84f4c9166256f81/ignite/engine/events.py#L144-L154

https://github.com/pytorch/ignite/blob/89d29303bcc11ce85ec5e010f84f4c9166256f81/ignite/engine/events.py#L96-L100

The test cases have also been written for them:
https://github.com/pytorch/ignite/blob/89d29303bcc11ce85ec5e010f84f4c9166256f81/tests/ignite/engine/test_custom_events.py#L316-L390

> I am not quite sure if I agree with you here, as in my opinion first=10 implies every iteration from 10 to +inf in the same way as after=9 does (and the same argument for last=20 <-> before=21). But as I said this might be personal preference, just wanted to state the idea

In my opinion, here `first=10` could also mean for the first 10 iterations 😅, so I think that would be a little confusing?

**Comment by vfdev-5** [comment]
> > we would like to pick events inside [0, before] or [after, inf]

@theory-in-progress this case does not work as we do 
```python
         if event > after_ and event < before_: 
             return True 
```

**Comment by vfdev-5** [comment]
Fixed by #2860 and #2858
